File: 0713. Subarray Product Less Than K.py
class Solution:
    def numSubarrayProductLessThanK(self, nums: List[int], k: int) -> int:
        L,ans = 0,0
        prod = 1
        for R,val in enumerate(nums):
            if val < k:
                ans += 1
            prod *= val
            while prod >= k and L < len(nums)-1:
                prod /= nums[L]
                L += 1
            if R - L > 0:
                ans += R - L
        return ans

File: 0713. Subarray Product Less Than K.js
// O(n) time and O(1) space complexity
var numSubarrayProductLessThanK = function(nums, k) {
    if (k == 0) { return 0 };

    let prod = 1, out = 0, L = 0;
    nums.map((num, R) => {

        if (num < k) { out += 1 };
        prod *= num

        if (prod >= k && R >= L) {
            while (prod >= k && L <= R) {
                prod /= nums[L];
                L++
            };
        };

        if (L <= R) {
            out += (R - L);
        };
    });

    return out;
};
